cpp/oneapi/dal/table/backend/homogen_table_impl.hpp

```diff
@@ -20,13 +20,84 @@
 
 namespace oneapi::dal::backend {
 
+enum class alloc_kind {
+    host, /// Non-USM pointer allocated on host
+    usm_host, /// USM pointer allocated by sycl::alloc_host
+    usm_device, /// USM pointer allocated by sycl::alloc_device
+    usm_shared /// USM pointer allocated by sycl::alloc_shared
+};
+
+#ifdef ONEDAL_DATA_PARALLEL
+inline alloc_kind alloc_kind_from_sycl(sycl::usm::alloc alloc) {
+    using error_msg = dal::detail::error_messages;
+    switch (alloc) {
+        case sycl::usm::alloc::host: return alloc_kind::usm_host;
+        case sycl::usm::alloc::device: return alloc_kind::usm_device;
+        case sycl::usm::alloc::shared: return alloc_kind::usm_shared;
+        default: throw invalid_argument{ error_msg::unsupported_usm_alloc() };
+    }
+}
+#endif
+
+#ifdef ONEDAL_DATA_PARALLEL
+inline sycl::usm::alloc alloc_kind_to_sycl(alloc_kind kind) {
+    switch (kind) {
+        case alloc_kind::usm_host: return sycl::usm::alloc::host;
+        case alloc_kind::usm_device: return sycl::usm::alloc::device;
+        case alloc_kind::usm_shared: return sycl::usm::alloc::shared;
+        case alloc_kind::host: return sycl::usm::alloc::unknown;
+        default: //
+            ONEDAL_ASSERT("Unsupported alloc_kind");
+    }
+}
+#endif
+
+inline bool alloc_kind_requires_copy(alloc_kind src_alloc_kind, alloc_kind dst_alloc_kind) {
+#ifdef ONEDAL_DATA_PARALLEL
+    switch (dst_alloc_kind) {
+        case alloc_kind::host: //
+            return (src_alloc_kind == alloc_kind::usm_device);
+        case alloc_kind::usm_host: //
+            return (src_alloc_kind == alloc_kind::host) || //
+                   (src_alloc_kind == alloc_kind::usm_device);
+        case alloc_kind::usm_device: //
+            return (src_alloc_kind == alloc_kind::host) || //
+                   (src_alloc_kind == alloc_kind::usm_host);
```

Comment:
Why it is **required** to copy in this case?

Reply:
This function determines efficient copy strategy. `case alloc_kind::usm_device:` corresponds to the following scenario:
```c++
auto x = accessor.pull(q, {0, -1}, sycl::usm::alloc::device);
``` 

If we do not copy, `x` will contain USM host memory. However, user specified USM device memory in `pull` and expect efficient access to `x` on device.

Reply:
I just trying to say that the name of the function does not clearly describe its intent. Yes, in terms of efficiency it is needed, but not required in terms of functional correctness

Reply:
Frankly speaking I don't like the current behavior. The problem is that if the user passes specific alloc kind, the function may return array allocated in different way.
```c++
auto x = accessor.pull(q, {0, -1}, sycl::usm::alloc::device);
// x could contain shared USM
```

The expected behavior to me would be if `x` is always allocated in the way that the user ask. However, it may require copy even if technically it's not needed, e.g., table contains shared but user ask for device.